diesel/src/connection/statement_cache.rs

```diff
@@ -111,3 +140,23 @@ impl<DB> StatementCacheKey<DB> where
         Ok(query_builder.finish())
     }
 }
+
+fn refmut_map_result<T, U, F>(refmut: RefMut<T>, f: F)
+    -> QueryResult<RefMut<U>> where
+        F: FnOnce(&mut T) -> QueryResult<&mut U>,
+{
+    use std::mem;
+
+    let mut error = None;
+    let result = RefMut::map(refmut, |mutref| match f(mutref) {
+        Ok(x) => x,
+        Err(e) => {
+            error = Some(e);
+            unsafe { mem::uninitialized() }
+        }
+    });
+    match error {
+        Some(e) => Err(e),
+        None => Ok(result),
+    }
+}
```

Comment:
I think there's a more fundamental issue with how types like `Option` and `Result` are able to compose. Adding more `try_` methods is a bandaid IMO. I think this is most apparent with deref coercions. If I have a `String` and want to call a method that takes `&str` I just write `&s`. If I have an `Option<String>` and want to call a method that takes an `Option<&str>` I have to write `s.as_ref().map(|x| &**x)`. Pattern matching also makes the issue quite apparent.